ql/src/java/org/apache/hadoop/hive/ql/optimizer/UnionDistinctMerger.java

```diff
@@ -0,0 +1,185 @@
+    public final ParseContext pCtx;
+
+    public UnionMergeContext(ParseContext pCtx) {
+      this.pCtx = pCtx;
+    }
+  }
+
+  private class UnionMergeProcessor implements SemanticNodeProcessor {
+    @Override
+    public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
+        Object... nodeOutputs) throws SemanticException {
+      UnionMergeContext context = (UnionMergeContext) procCtx;
+
+      // The stack contains at least 8 operators, UNION-GBY-RS-GBY-UNION-GBY-RS-GBY.
+      // The leftmost UNION is on stack.size() - 8 and the rightmost GBY is on stack.size() - 1.
+      Set<Operator> allOps = new HashSet<>(context.pCtx.getAllOps());
+      for (int i = 1; i <= 8; i ++) {
+        Operator<?> op = (Operator<?>) stack.get(stack.size() - i);
+
+        // We do not apply the optimization if some operators do not belong to query plan.
+        // This can be happened when we already merged some UNIONs before.
+        // For example, Suppose that a query plan looks like the below graph:
+        //   (1)UNION-GBY-RS-GBY-(3)UNION-GBY-RS-GBY
+        //   (2)UNION-GBY-RS-GBY-
+        // Then we merge (1)-(3) to (1) and them move on to merging (2)-(3). Without checking the presence of
+        // operators in (2) and (3), the merge process will fail as (3) is already removed.
+        if (!allOps.contains(op)) {
+          return null;
+        }
+
+        // We do not apply the optimization if intermediate outputs are used by other operators.
+        if (i != 1 && op.getChildOperators().size() > 1) {
+          return null;
+        }
+      }
+
+      UnionOperator upperUnionOperator = (UnionOperator) stack.get(stack.size() - 8);
+      GroupByOperator upperFinalGroupByOperator = (GroupByOperator) stack.get(stack.size() - 5);
+
+      UnionOperator lowerUnionOperator = (UnionOperator) stack.get(stack.size() - 4);
+      GroupByOperator lowerFinalGroupByOperator = (GroupByOperator) stack.get(stack.size() - 1);
+
+      // We can apply the optimization if there is no aggregators in final GroupBy operators. The absence of
+      // aggregators ensures that we are merging two distinct computation.
+      if (upperFinalGroupByOperator.getConf().getAggregators().isEmpty() &&
+          lowerFinalGroupByOperator.getConf().getAggregators().isEmpty()) {
+        LOG.info("Detect duplicate UNION-DISTINCT GBY patterns. Remove the latter one.");
+
+        // Step 0. UNION1->GBY1->RS1->GBY2->UNION2->GBY3->RS2->GBY4
+
+        // Step 1. Cut GBY2->UNION2
+        lowerUnionOperator.removeParent(upperFinalGroupByOperator);
+
+        // Step 2.
+        //   Connect the parent of lowerUnionOperator and upperUnionOperator.
+        //   Disconnect lowerUnionOperator from operator graph.
+        // Before step 2:
+        //    {OP1, 2}-UNION1->GBY1->RS1->GBY2-{}
+        //    {OP3, 4}-UNION2->GBY3->RS2->GBY4-{OP5, 6, ...}
+        // After step 2:
+        //    {OP1, 2}-UNION1->GBY1->RS1->GBY2-{}
```

Comment:
`{OP1, 2, 3, 4}-UNION1->GBY1->RS1->GBY2-{}` or `{OP1, 2, ...}-UNION1->GBY1->RS1->GBY2-{}` If I understand correctly?

Reply:
missed that. @ngsg could you please check

Reply:
It should be {OP1, 2, 3, 4}. Thanks for the thorough review.